src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,141 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static class Node<T> {
+        private T element;
+        private Node<T> next;
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        indexCheck(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> previous = getNodeByIndex(index - 1);
+            newNode.next = previous.next;
+            previous.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Unable to set index: " + index);
+        }
+        T removedElement;
+        if (index == 0) {
+            removedElement = head.element;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else if (index == size - 1) {
+            Node<T> previous = getNodeByIndex(index - 1);
+            removedElement = tail.element;
+            previous.next = null;
+            tail = previous;
+        } else {
+            Node<T> previous = getNodeByIndex(index - 1);
+            Node<T> current = previous.next;
+            removedElement = current.element;
+            previous.next = current.next;
+        }
+        size--;
+        return removedElement;
```

Comment:
See common mistakes, you should implement method unlink and reuse it in remove methods. Just move this code in new method
```suggestion
        return  unlink(getNodeByIndex(index));
```